src/yggdrasil/tcp.go

```diff
@@ -62,104 +56,145 @@ func (iface *tcpInterface) setExtraOptions(c net.Conn) {
+			if len(added) > 0 || len(deleted) > 0 {
+				for _, a := range added {
+					if a[:6] != "tcp://" {
+						continue
+					}
+					if _, err := t.listen(a[6:]); err != nil {
+						e <- err
+						continue
+					}
+				}
+				for _, d := range deleted {
+					if d[:6] != "tcp://" {
+						continue
+					}
+					t.mutex.Lock()
+					if listener, ok := t.listeners[d[6:]]; ok {
+						t.mutex.Unlock()
+						listener.stop <- true
+					} else {
+						t.mutex.Unlock()
+					}
+				}
+				e <- nil
 			} else {
 				e <- nil
 			}
 		}
 	}()
 
-	return iface.listen()
+	t.link.core.configMutex.RLock()
+	defer t.link.core.configMutex.RUnlock()
+	for _, listenaddr := range t.link.core.config.Listen {
+		if listenaddr[:6] != "tcp://" {
+			continue
+		}
+		if _, err := t.listen(listenaddr[6:]); err != nil {
+			return err
+		}
+	}
+
+	return nil
 }
 
-func (iface *tcpInterface) listen() error {
+func (t *tcp) listen(listenaddr string) (*tcpListener, error) {
 	var err error
 
-	iface.core.configMutex.RLock()
-	iface.addr = iface.core.config.Listen
-	iface.core.configMutex.RUnlock()
-
 	ctx := context.Background()
 	lc := net.ListenConfig{
-		Control: iface.tcpContext,
+		Control: t.tcpContext,
 	}
-	iface.serv, err = lc.Listen(ctx, "tcp", iface.addr)
+	listener, err := lc.Listen(ctx, "tcp", listenaddr)
 	if err == nil {
-		iface.mutex.Lock()
-		iface.calls = make(map[string]struct{})
-		iface.conns = make(map[tcpInfo](chan struct{}))
-		iface.mutex.Unlock()
-		go iface.listener()
-		return nil
+		l := tcpListener{
+			listener: listener,
+			stop:     make(chan bool),
+		}
+		go t.listener(&l, listenaddr)
+		return &l, nil
 	}
 
-	return err
+	return nil, err
 }
 
 // Runs the listener, which spawns off goroutines for incoming connections.
-func (iface *tcpInterface) listener() {
-	defer iface.serv.Close()
-	iface.core.log.Infoln("Listening for TCP on:", iface.serv.Addr().String())
+func (t *tcp) listener(l *tcpListener, listenaddr string) {
+	if l == nil {
+		return
+	}
+	// Track the listener so that we can find it again in future
+	t.mutex.Lock()
+	t.listeners[listenaddr] = l
+	t.mutex.Unlock()
+	// And here we go!
+	accepted := make(chan bool)
+	defer l.listener.Close()
+	t.link.core.log.Infoln("Listening for TCP on:", l.listener.Addr().String())
 	for {
-		sock, err := iface.serv.Accept()
-		if err != nil {
-			iface.core.log.Errorln("Failed to accept connection:", err)
-			return
-		}
+		var sock net.Conn
+		var err error
+		// Listen in a separate goroutine, as that way it does not block us from
+		// receiving "stop" events
+		go func() {
+			sock, err = l.listener.Accept()
+			accepted <- true
+		}()
+		// Wait for either an accepted connection, or a message telling us to stop
+		// the TCP listener
 		select {
-		case <-iface.stop:
-			iface.core.log.Errorln("Stopping listener")
-			return
-		default:
+		case <-accepted:
 			if err != nil {
-				panic(err)
+				t.link.core.log.Errorln("Failed to accept connection:", err)
+				return
```

Comment:
We should do the same cleanup (from the `l.stop` case) if we return for any reason. I guess maybe move that to a `defer` right below this loop starts, and then both `l.stop` and this `err` case could be a line or so of logging + a `return`, unless you have a better idea.